### ISSUE TITLE
Prompt for User Credentials for Implicit/Authorization Code OAuth 2.0 Authentication

### ISSUE THREAD

**Issue by pankajdhakre** [strongloop/loopback-gateway#21]
Hi,

I tried to follow up the steps and setup the demo as mentioned.
But after logging to account using username/password as bob/secret, I am trying to authenticate using any of the following type 
Implicit
Authorization Code
Client Credentials or Resource Owner Password Credentials
It redirects me and prompts for username and password. I tried username as 'bob' and password as 'secret', but I am getting the JSON response as
{"error":"invalid_client","error_description":"Invalid client"}.

Attaching the snapshot for the same.

![screenshot from 2015-04-22 14 01 55](https://cloud.githubusercontent.com/assets/11556186/7270479/115e1200-e8fa-11e4-9c89-c4913d92b810.png)
![screenshot from 2015-04-22 14 02 02](https://cloud.githubusercontent.com/assets/11556186/7270480/1162c138-e8fa-11e4-8791-b73d18cd2d75.png)
![screenshot from 2015-04-22 14 02 08](https://cloud.githubusercontent.com/assets/11556186/7270481/1164837e-e8fa-11e4-96c1-9f754f9dbd36.png)

Please let me know what is the issue or I am missing any step..

**Comment by raymondfeng** [comment]
The prompt is for client_id/client_secret, please try '123' and 'secret'.

**Comment by pankajdhakre** [comment]
Hi raymondfeng,
I tried it for client_id/client_secret, please try '123' and 'secret', but its not working.
Also in prompt is asks for username/password not for client_id/client_secret.

Is it somehow related to keys and certificates provided in ./server/private folder.?

**Comment by Spatlani** [comment]
having the same issue, the prompt for username and password keeps on coming even after they're entered.

If I press Esc than the invalid client error occurs

**Comment by PanYuntao** [comment]
I encountered the same problem.

**Comment by woodedlawn** [comment]
I was banging my head on this one too, but I think I've figured it out.

The change to loopback-component-oauth2/lib/models/oauth2-models.js in loopback-component-oauth2 v2.0.0-beta8 caused the 'OAuthClientApplication' model to inherit from the built-in 'Application' model. This sets 'OAuthClientApplication' as the 'applicationModel' in lib/models/index.js because it looks up the model by type which finds models configured over base models.

The [signupTestUserAndApp function](https://github.com/strongloop/loopback-gateway/blob/master/server/server.js#L107-L142) registers an 'Application', but it should be registering an 'OAuthClientApplication'. I made the following changes and it works for me now.

```javascript
function signupTestUserAndApp() {
// Create a dummy user and client app
  app.models.User.create({username: 'bob',
    password: 'secret',
    email: 'foo@bar.com'}, function(err, user) {

    if (!err) {
      console.log('User registered: username=%s password=%s',
        user.username, 'secret');
    }

    // Hack to set the app id to a fixed value so that we don't have to change
    // the client settings
    var OAuthClientApplication = app.dataSources.db.getModel("OAuthClientApplication");
    OAuthClientApplication.beforeSave = function(next) {
      this.id = 123;
      this.restApiKey = 'secret';
      next();
    };
    OAuthClientApplication.register(
      user.id,
      'demo-app',
      {
        publicKey: sslCert.certificate
      },
      function(err, demo) {
        if (err) {
          console.error(err);
        } else {
          console.log('Client application registered: id=%s key=%s',
            demo.id, demo.restApiKey);
        }
      }
    );

  });
}
```

**Comment by raymondfeng** [comment]
@woodedlawn Good catch!

**Comment by raymondfeng** [comment]
Fixed and released in 1.1.1
